encoders/firebase-encoders/src/main/java/com/google/firebase/encoders/ObjectEncoderContext.java

```diff
@@ -68,4 +68,28 @@ ObjectEncoderContext add(@NonNull String name, double value)
   @NonNull
   ObjectEncoderContext add(@NonNull String name, boolean value)
       throws IOException, EncodingException;
+
+  /**
+   * Begin a nested JSON object.
+   *
+   * <p>Unlike {@code add()} methods, this method returns a new "child" context that's used to
+   * populate the nested JSON object. This context can only be used until the parent context is
+   * mutated by calls to {@code add()} or {@code nested()}, violating this will result in a {@link
+   * IllegalStateException}.
+   *
+   * <p>Nesting can be arbitrarily deep.
+   *
+   * <p>Example:
+   *
+   * <pre>{@code
+   * ctx.add("key", "value");
+   * ObjectEncoderContext nested = ctx.nested("nested");
+   * nested.add("key", "value");
+   *
+   * // After this call the above nested context is invalid.
+   * ctx.add("anotherKey", 1);
+   * }</pre>
+   */
+  @NonNull
+  ObjectEncoderContext nested(@NonNull String name) throws IOException;
```

Comment:
if `nested` received a "runnable" as second parameter we could do something like

```
ctx.add("key", "value")
   .nested("key", (nctx) -> {
        nctx.add("nested_key", "nested_value")})
    .add("second_key", "second_value"));
````
we could get rid of keeping track of whether there's a nested context to close or not. wdyt?

Reply:
I generally like your interface much better as it makes it "impossible to use the type incorrectly (c)".

However I think we considered it at some point but the problem with lambdas is that they will result in class bloat which will count towards dex limit(as each lambda is effectively an anonymous class), whereas returning a nested context does not have this problem, so no matter how much I "nest" I end up with just one class(encoder that I am implementing itself). I am not certain it is a big deal though

Reply:
You're right, I forgot we discussed this earlier. Let's run with this and see how it goes :-).